### PR TITLE
more robust service creation

### PR DESCRIPTION
**Pull request by classicist**
Hi,
    Thank you for writing this great gem. I needed to extend the RPC functionality for my distributed team and thought the code might be more generally useful. The changes I made only extend the functionally you had coded already, so it will not break backward compatibility. Please let me know if you think these changes would be more generally useful. Thanks again for your excellent work -- yours is the clear winner as far as protobuf gems go.

Commit Msg: added more robust service creation with type-safe responses, injected service handlers, and a service registry to faciltate pure-ruby RPC without coupling it to any particular communication strategy (TCPSocket, RabbitMQ, etc) or application architecture.

### CODE REVIEW

File: lib/protocol_buffers.rb [hunk 1]
```diff
@@ -4,13 +4,15 @@ module ProtocolBuffers
   # for 1.9.2 compatibility
   def self.bin_sio(*args)
     sio = StringIO.new(*args)
-    sio.set_encoding('binary') if
```
Comment (bufdev) [inline lib/protocol_buffers.rb:7]: whitespace

File: lib/protocol_buffers.rb [hunk 2]
```diff
@@ -4,13 +4,15 @@ module ProtocolBuffers
...
     sio = StringIO.new(*args)
-    sio.set_encoding('binary') if 
+    sio.set_encoding('binary') if
       sio.respond_to?(:set_encoding) and
       sio.external_encoding != Encoding::ASCII_8BIT
     sio
   end
 end
 
+$LOAD_PATH << File.dirname(__FILE__)
```
Comment (bufdev) [inline lib/protocol_buffers.rb]: whitespace

File: lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb [hunk 3]
```diff
@@ -46,7 +47,7 @@ def write(io)
         dump_service(descriptor.package, service)
       end
     end
-
```
Comment (bufdev) [inline lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb:49]: whitespace

File: lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb [hunk 4]
```diff
@@ -77,11 +78,11 @@ def line(str = nil)
     end
     @io.write("\n")
   end
-
```
Comment (bufdev) [inline lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb:80]: whitespace

File: lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb [hunk 5]
```diff
@@ -77,11 +78,11 @@ def line(str = nil)
     end
     @io.write("\n")
   end
-  
+
   def in_namespace(type, namespace, rest = "")
-
```
Comment (bufdev) [inline lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb:82]: whitespace

File: lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb [hunk 6]
```diff
@@ -170,13 +171,21 @@ def dump_enum(package, enum)
     line
   end
 
-  def dump_service(package, service)
```
Comment (bufdev) [inline lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb:173]: is this correct indentation?

File: lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb [hunk 7]
```diff
@@ -170,13 +171,21 @@ def dump_enum(package, enum)
...
+def dump_service(package, service)
     in_namespace("class", service.name, " < ::ProtocolBuffers::Service") do
       fully_qualified_name = fully_qualified_name(package, service.name)
       line %{set_fully_qualified_name "#{fully_qualified_name}"}
       line
       service.method.each do |method|
         line %{rpc :#{underscore(method.name)}, "#{method.name}", #{service_typename(method.input_type)}, #{service_typename(method.output_type)}}
+        unless fully_qualified_name.nil?
```
Comment (bufdev) [inline lib/protocol_buffers/compiler/file_descriptor_to_ruby.rb:181]: silent error

File: lib/protocol_buffers/compiler/fully_qualified_name.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
+module ProtocolBuffers
+  class FullyQualifiedName
+    def self.to_class(fully_qualified_name)
+      return nil if fully_qualified_name.nil?
+      service_typename(fully_qualified_name.to_s).split('::').inject(Object) { |mod, class_name| mod.const_get(class_name) if mod.const_defined?(class_name)}
+    end
+
+    def self.service_typename(type_name)
```
Comment (bufdev) [inline lib/protocol_buffers/compiler/fully_qualified_name.rb:11]: I dont think this kind of stuff should be exposed outside the compiler

File: lib/protocol_buffers/compiler/fully_qualified_name.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      type_name.split(".").map { |t| camelize(t) }.join("::")
+    end
+
+    def self.camelize(lower_case_and_underscored_word)
+      lower_case_and_underscored_word.to_s.gsub(/(?:^|_)(.)/) { $1.upcase }
+    end
+
+    def self.underscore(camelized_word)
```
Comment (bufdev) [inline lib/protocol_buffers/compiler/fully_qualified_name.rb:19]: Same comment on exposing outside the compiler

File: lib/protocol_buffers/compiler/fully_qualified_name.rb [hunk 10]
```diff
@@ -0,0 +1,35 @@
+module ProtocolBuffers
+  class FullyQualifiedName
+    def self.to_class(fully_qualified_name)
+      return nil if fully_qualified_name.nil?
```
Comment (bufdev) [inline lib/protocol_buffers/compiler/fully_qualified_name.rb:4]: silent error

File: lib/protocol_buffers/runtime/message.rb [hunk 11]
```diff
@@ -231,7 +231,8 @@ class Message
     #   # is equivalent to
     #   message = MyMessageClass.new
     #   message.attributes = attributes
-    def initialize(attributes = {})
+    def initialize(attributes = nil)
```
Comment (bufdev) [inline lib/protocol_buffers/runtime/message.rb:234]: revert

File: lib/protocol_buffers/runtime/message.rb [hunk 12]
```diff
@@ -513,7 +514,7 @@ def self.valid?(message, raise_exception=false)
         next if field.otype != :required
         next if message.value_for_tag?(tag) && (field.class != Field::MessageField || message.value_for_tag(tag).valid?)
         return false unless raise_exception
-        raise(ProtocolBuffers::EncodeError.new(field), "Required field '#{field.name}' is invalid")
+        raise(ProtocolBuffers::EncodeError.new(field), "Required field '#{field.name}' for '#{self}' is invalid")
```
Comment (bufdev) [inline lib/protocol_buffers/runtime/message.rb:547]: revert

File: lib/protocol_buffers/runtime/message.rb [hunk 13]
```diff
@@ -338,7 +339,7 @@ def ==(obj)
       end
       return true
     end
-
```
Comment (bufdev) [inline lib/protocol_buffers/runtime/message.rb]: whitespace

File: lib/protocol_buffers/runtime/service.rb [hunk 14]
```diff
@@ -1,10 +1,60 @@
 require 'protocol_buffers/runtime/rpc'
+require 'protocol_buffers/compiler/file_descriptor_to_ruby'
+require 'protocol_buffers/compiler/fully_qualified_name'
+
+module ProtocolBuffers
+  class ServiceRegistry
```
Comment (bufdev) [inline lib/protocol_buffers/runtime/service.rb:6]: document if this is going to stay

File: lib/protocol_buffers/runtime/service.rb [hunk 15]
```diff
@@ -1,10 +1,60 @@
...
+    def self.register_service(fully_qualified_name, handler_factory)
+      return nil unless fully_qualified_name && handler_factory
```
Comment (bufdev) [inline lib/protocol_buffers/runtime/service.rb:8]: silent error on fully_qualified_name

File: lib/protocol_buffers/runtime/service.rb [hunk 16]
```diff
@@ -1,10 +1,60 @@
...
+      @handler_factory_registry ||= Hash.new
```
Comment (bufdev) [inline lib/protocol_buffers/runtime/service.rb:9]: In your tests, you don't subclass ServiceRegistry, which effectively means you are treating this as a singleton. This means if someone does subclass this, your instance variables (@) will result in two different set of values for @handler_factory_registry.

This is also not thread-safe but I don't think that's a requirement overall.

File: lib/protocol_buffers/runtime/service.rb [hunk 17]
```diff
@@ -1,10 +1,60 @@
...
+    def self.create_service(handler)
+      new(handler)
+    end
+
+    def ensure_handler_defined!
+      raise "No handler defined for #{self}\n #{Exception.new.backtrace}" unless @handler
+    end
+
+    def ensure_correct_request_type!(method_name, message)
+      raise "Request Type Error: #{message} must be an instance of #{types_for(method_name)[:request]}" unless message.instance_of?(types_for(method_name)[:request])
```
Comment (bufdev) [inline lib/protocol_buffers/runtime/service.rb:55]: specify what type of StandardError you want to raise?

File: lib/protocol_buffers/runtime/service.rb [hunk 18]
```diff
@@ -1,10 +1,60 @@
...
+      (@handler_factory_registry || {}).keys
+    end
+
+    def self.to_class(fully_qualified_name)
+      ProtocolBuffers::FullyQualifiedName.to_class(fully_qualified_name)
+    end
+  end
+end
 
 module ProtocolBuffers
```
Comment (bufdev) [inline lib/protocol_buffers/runtime/service.rb:37]: This needs documentation now that it is more complicated.

File: spec/proto_files/services.pb.rb [hunk 19]
```diff
@@ -38,13 +38,41 @@ class FooBarService < ::ProtocolBuffers::Service
...
+    def get_bar(message)
+      ensure_handler_defined!
+      ensure_correct_request_type!(:get_bar, message)
+      raw_reponse = @handler.send(:get_bar, message.to_hash)
+      types_for(:get_bar)[:response].new(raw_reponse)
+    end
+
   end
   class NoNameFooBarService < ::ProtocolBuffers::Service
-    # purposefully removing qualified name to make sure that nothing breaks
```
Comment (bufdev) [inline spec/proto_files/services.pb.rb:44]: whitepace

File: spec/proto_files/services.pb.rb [hunk 20]
```diff
@@ -38,13 +38,41 @@ class FooBarService < ::ProtocolBuffers::Service
...
     rpc :get_bar, "GetBar", ::Services::BarRequest, ::Services::BarResponse
+    def get_bar(message)
+      ensure_handler_defined!
+      ensure_correct_request_type!(:get_bar, message)
+      raw_reponse = @handler.send(:get_bar, message.to_hash)
+      types_for(:get_bar)[:response].new(raw_reponse)
+    end
+
   end
-end
```
Comment (bufdev) [inline spec/proto_files/services.pb.rb:50]: whitespace

**Comment by codekitchen** [comment]
Hi @emschwar , @jgeiger are you looking to revive this PR? I'm not really sure what to make of your recent pushes, but it's been years so it'd take me a while to refresh my cache on what this is all about. Please let me know if you plan to pursue this. Thanks.

**Comment by codekitchen** [comment]
closing at request of @jgeiger-trane

**Pull request closed without merging**